bd-log-primitives/src/lib.rs

```diff
@@ -215,38 +252,239 @@ pub enum LogFieldKind {
+  pub fn field_value<'a>(&'a self, field_key: &str) -> Option<Cow<'a, str>> {
+    FieldsRef::new(&self.fields, &self.matching_fields).field_value(field_key)
+  }
 
-/// A reference to a log message and its associated fields.
-#[derive(Clone, Copy, Debug)]
-pub struct LogRef<'a> {
-  pub log_type: LogType,
-  pub log_level: LogLevel,
-  pub message: &'a LogMessage,
-  pub fields: FieldsRef<'a>,
-  pub session_id: &'a str,
-  pub occurred_at: time::OffsetDateTime,
+  fn size_proto_data(value: &LogFieldValue) -> u64 {
+    let mut my_size = 0;
+    match value {
+      StringOrBytes::String(s) => {
+        my_size += ::protobuf::rt::string_size(1, s);
+      },
+      StringOrBytes::SharedString(s) => {
+        my_size += ::protobuf::rt::string_size(1, s.as_ref());
+      },
+      StringOrBytes::Bytes(b) => {
+        // This encodes the Binary proto message.
+        let inner_len = ::protobuf::rt::bytes_size(2, b);
+        my_size += 1 + ::protobuf::rt::compute_raw_varint64_size(inner_len) + inner_len;
+      },
+    }
+
+    my_size
+  }
+
+  fn serialize_proto_data(
+    field_number: u32,
+    value: &LogFieldValue,
+    os: &mut CodedOutputStream<'_>,
+  ) -> anyhow::Result<()> {
+    os.write_tag(field_number, WireType::LengthDelimited)?;
+    os.write_raw_varint32(Self::size_proto_data(value).to_u32())?;
+    match value {
+      StringOrBytes::String(s) => {
+        os.write_string(1, s)?;
+      },
+      StringOrBytes::SharedString(s) => {
+        os.write_string(1, s.as_ref())?;
+      },
+      StringOrBytes::Bytes(b) => {
+        // This encodes the Binary proto message.
+        let inner_len = ::protobuf::rt::bytes_size(2, b);
+        os.write_tag(2, WireType::LengthDelimited)?;
+        os.write_raw_varint32(inner_len.to_u32())?;
+        os.write_bytes(2, b)?;
+      },
+    }
+
+    Ok(())
+  }
+
+  fn size_proto_field(key: &LogFieldKey, value: &LogFieldValue) -> u64 {
+    let mut my_size = 0;
+    my_size += ::protobuf::rt::string_size(1, key.as_ref());
+    let value_len = Self::size_proto_data(value);
+    my_size += 1 + ::protobuf::rt::compute_raw_varint64_size(value_len) + value_len;
+    my_size
+  }
+
+  fn serialize_proto_field(
+    field_number: u32,
+    key: &LogFieldKey,
+    value: &LogFieldValue,
+    os: &mut CodedOutputStream<'_>,
+  ) -> anyhow::Result<()> {
+    os.write_tag(field_number, WireType::LengthDelimited)?;
+    os.write_raw_varint32(Self::size_proto_field(key, value).to_u32())?;
+    os.write_string(1, key.as_ref())?;
+    Self::serialize_proto_data(2, value, os)?;
+    Ok(())
+  }
+
+  #[must_use]
+  pub fn serialize_proto_size_inner(
+    log_level: u32,
+    message: &LogFieldValue,
+    fields: &LogFields,
+    session_id: &str,
+    occurred_at: OffsetDateTime,
+    log_type: LogType,
+    action_ids: &[&str],
+    stream_ids: &[&str],
+  ) -> u64 {
+    let mut my_size = 0;
+
+    my_size += ::protobuf::rt::uint64_size(1, occurred_at.unix_timestamp_nanos().to_u64() / 1000);
```

Comment:
where does nanos / 1000 come from?

Reply:
It's microseconds since epoch. I will add more comments everywhere.